presto-druid/src/main/java/com/facebook/presto/druid/DruidClient.java

```diff
@@ -83,6 +102,46 @@ public DruidClient(DruidConfig config, @ForDruidClient HttpClient httpClient)
         this.druidCoordinator = URI.create(config.getDruidCoordinatorUrl());
         this.druidBroker = URI.create(config.getDruidBrokerUrl());
         this.druidSchema = config.getDruidSchema();
+        this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+
+        Duration caseInsensitiveNameMatchingCacheTtl = requireNonNull(config.getCaseInsensitiveNameMatchingCacheTtl(), "caseInsensitiveNameMatchingCacheTtl is null");
+        CacheBuilder<Object, Object> remoteTableNamesCacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteTables = remoteTableNamesCacheBuilder.build();
+    }
+
+    Optional<RemoteTableObject> toRemoteTable(SchemaTableName schemaTableName)
+    {
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(schemaTableName.getTableName()), "Expected table name from internal metadata to be lowercase: %s", schemaTableName);
+        if (!caseInsensitiveNameMatching) {
+            return Optional.of(RemoteTableObject.of(schemaTableName.getTableName()));
+        }
+
+        @Nullable Optional<RemoteTableObject> remoteTable = remoteTables.getIfPresent(schemaTableName);
+        if (remoteTable != null) {
+            return remoteTable;
+        }
+
+        // Cache miss, reload the cache
+        Map<SchemaTableName, Optional<RemoteTableObject>> mapping = new HashMap<>();
+        for (String table : getTables()) {
+            SchemaTableName cacheKey = new SchemaTableName(getSchema(), table);
+            mapping.merge(
+                    cacheKey,
+                    Optional.of(RemoteTableObject.of(table)),
```

Comment:
what's the benefit of using an Optional here? can we just put the remoteTableObject as the value directly?

Reply:
I am using Optional.empty() to mark a table which doesn't exist in Druid.

Reply:
Ok, it make sense.  Thank you for the clarification!